File: qbopt/optimize/loopmotion.py
from dataclasses import replace
from functools import cache

from qbopt.analysis import loops
from qbopt.model import mir
from qbopt.analysis import induction
from qbopt.analysis import ranges
from qbopt.analysis import consts
from qbopt.objectfile.module import Space


def sunk_stores(body: mir.MirBody, dgroup: frozenset[int], bounds: dict | None = None) -> mir.MirBody:
    predecessors = loops.predecessors(body.blocks)
    for loop in loops.loops(body.blocks, body.entry):
        scoped = ranges.bounded(body)
        # With constants, as hoist asks: a store through the literal selector
        # 0A000h otherwise observes every frame and descriptor cell.
        constant = ranges.constants(body, dgroup)
        intervals = {id(op): {**constant, **scoped.get(block.at, {})} for block in body.blocks for op in block.ops}
        blocks = {block.at: block for block in body.blocks}
        inside = [blocks[at] for at in loop.body]
        exits = {(block.at, to) for block in inside for to in block.succ if to not in loop.body}
        if len(exits) != 1:
            continue
        source, destination = next(iter(exits))
        if destination not in blocks or predecessors[destination] != frozenset({source}):
            continue
        if any(not block.succ for block in inside):
            continue
        operations = [op for block in inside for op in block.ops]
        if any(
            op.floating is not None or op.barrier
            or op.kind in {mir.Kind.CALL, mir.Kind.ARG, mir.Kind.OPAQUE, mir.Kind.ESCAPE, mir.Kind.RETURN}
            for op in operations
        ):
            continue
        exit_block = blocks[destination]
        if not exit_block.ops:
            continue
        dominators = loops.dominators(body.blocks, body.entry)
        address_values = {value for block in body.blocks
                          if block.at not in loop.body and block.at in dominators.get(loop.header, ())
                          for value in (*[phi.result for phi in block.phis],
                                        *[value for op in block.ops for value in op.defines])}
        moved = [op for op in blocks[source].ops
                 if _unobserved(op, operations, dgroup, bounds, intervals, address_values)]
        relocated = {id(op): op for op in moved}
        if len(loop.latches) == 1 and source == loop.header:
            latch = next(iter(loop.latches))
            outside = predecessors[source] - loop.body
            if len(outside) == 1 and latch != source and blocks[latch].succ == (source,):
                entry = next(iter(outside))
                nonempty = induction.nonempty(body, loop)
                invariant = induction.invariant(body, set(loop.body)) if nonempty else set()
                for op in operations:
                    if id(op) not in relocated and _unobserved(op, operations, dgroup, bounds, intervals, address_values):
                        value = _exit_value(op, blocks[source], blocks[entry], latch, body, dgroup, bounds)
                        if value is None and any(op is one for one in blocks[latch].ops):
                            value = _invariant_value(op, invariant, nonempty)
                            if value is None:
                                value = _last_counter_value(op, body, loop)
                        if value is not None:
                            moved.append(op)
                            relocated[id(op)] = replace(
                                op,
                                args=tuple(value if isinstance(arg, mir.Held) else arg for arg in op.args),
                                uses=tuple(dict.fromkeys(([value.value] if isinstance(value, mir.Held) else [])
                                    + [part for ref in op.stores for part in (ref.base, ref.segment) if part is not None])),
                            )
        if not moved:
            continue
        identities = {id(op) for op in moved}
        updates = {
            **{
                block.at: replace(block, ops=tuple(op for op in block.ops if id(op) not in identities))
                for block in inside
            },
            destination: replace(
                exit_block,
                ops=tuple(replace(relocated[id(op)], at=exit_block.ops[0].at) for op in moved) + exit_block.ops,
            ),
        }
        body = replace(body, blocks=tuple(updates.get(block.at, block) for block in body.blocks))
    return body


def _last_counter_value(op: mir.Op, body: mir.MirBody, loop) -> mir.Const | None:
    if len(op.args) != 1 or not isinstance(op.args[0], mir.Held):
        return None
    stored = op.args[0]
    counter = induction.basics(body, loop).get(stored.value.id)
    if counter is None or counter.start.width != stored.width or op.stores[0].width != stored.width:
        return None
    last = induction._last_counter(body, loop, counter, consts.known(body), stored.width)
    return mir.Const(consts.masked(last, stored.width), stored.width) if last is not None else None


def _invariant_value(op: mir.Op, invariant: set[int], nonempty: bool) -> mir.Arg | None:
    if not nonempty or len(op.args) != 1:
        return None
    match op.args[0]:
        case mir.Const() as value:
            return value
        case mir.Held(value=value) as arg if value.id in invariant:
            return arg
    return None


def _exit_value(
    op: mir.Op,
    header: mir.MirBlock,
    entry: mir.MirBlock,
    latch: int,
    body: mir.MirBody,
    dgroup: frozenset[int],
    bounds: dict | None,
) -> mir.Held | None:
    values = [arg for arg in op.args if isinstance(arg, mir.Held)]
    if len(values) != 1:
        return None
    stored = values[0]
    definitions = {value: operation for block in body.blocks for operation in block.ops for value in operation.defines}

    def root(arg: mir.Arg) -> mir.Arg:
        seen = set()
        while isinstance(arg, mir.Held) and arg.value not in seen:
            seen.add(arg.value)
            defining = definitions.get(arg.value)
            if defining is None or defining.kind is not mir.Kind.COPY or len(defining.args) != 1:
                break
            if not any(
                isinstance(result, mir.Held) and result.value == arg.value and result.width >= arg.width
                for result in defining.results
            ):
                break
            source = defining.args[0]
            if not isinstance(source, (mir.Held, mir.Const)) or source.width < arg.width:
                break
            arg = (
                mir.Held(source.value, arg.width)
                if isinstance(source, mir.Held)
                else mir.Const(source.n & ((1 << (arg.width * 8)) - 1), arg.width)
            )
        return arg

    ref = op.stores[0]
    blocks = {block.at: block for block in body.blocks}
    predecessors = loops.predecessors(body.blocks)

    @cache
    def memory():
        barriers = {one.at: "" for block in body.blocks for one in block.ops
                    if one.barrier or one.kind in {mir.Kind.CALL, mir.Kind.ESCAPE, mir.Kind.OPAQUE}}
        return consts.cells(body, dgroup, barriers), barriers

    def stored_at(at: int, expected: mir.Arg, active: frozenset) -> bool:
        expected = root(expected)
        key = (at, expected)
        if key in active:
            return True  # inductive backedge; every entry path still needs a matching store
        block = blocks[at]
        for index in range(len(block.ops) - 1, -1, -1):
            previous = block.ops[index]
            if previous.barrier or previous.kind in {mir.Kind.CALL, mir.Kind.ESCAPE, mir.Kind.OPAQUE}:
                return False
            if any(mir.overlapping(ref, written, dgroup, bounds) for written in previous.stores):
                if isinstance(expected, mir.Const):
                    fact = consts.initialized(previous, ref)
                    if fact is None:
                        facts, barriers = memory()
                        after = consts._kills(facts.get((at, index), {}), previous, {}, dgroup, barriers)
                        fact = consts._cell(after, ref)
                    if fact == consts.Known(consts.masked(expected.n, expected.width), expected.width):
                        return True
                args = [arg for arg in previous.args if isinstance(arg, (mir.Const, mir.Held))]
                return (
                    previous.kind is mir.Kind.STORE
                    and previous.stores == (ref,)
                    and len(args) == 1
                    and root(args[0]) == expected
                )
        if at == body.entry or not predecessors[at]:
            return False
        phi = next((phi for phi in block.phis if isinstance(expected, mir.Held) and phi.result == expected.value), None)
        if phi is not None and set(phi.incoming) != predecessors[at]:
            return False
        return all(
            stored_at(
                parent,
                mir.Held(phi.incoming[parent], expected.width) if phi is not None else expected,
                active | {key},
            )
            for parent in predecessors[at]
        )

    for phi in header.phis:
        if set(phi.incoming) != {entry.at, latch}:
            continue
        seed = root(mir.Held(phi.incoming[entry.at], stored.width))
        if stored_at(entry.at, seed, frozenset()) and stored_at(latch, mir.Held(phi.incoming[latch], stored.width), frozenset()):
            return mir.Held(phi.result, stored.width)
    return None


def _unobserved(op: mir.Op, operations: list[mir.Op], dgroup: frozenset[int], bounds: dict | None,
                intervals: dict | None = None, address_values: set | frozenset = frozenset()) -> bool:
    if op.kind is not mir.Kind.STORE or op.loads or op.defines or len(op.stores) != 1:
        return False
    ref = op.stores[0]
    if ref.addr is None or ref.addr.space not in (Space.SEGMENT, Space.FRAME) or ref.segment is not None:
        return False
    if ref.base is not None and (ref.base not in address_values or not ref.excludes):
        return False
    return not any(
        mir.overlapping(ref, other, dgroup, bounds, known=(intervals or {}).get(id(op)),
                        other_known=(intervals or {}).get(id(one)))
        for one in operations
        if one is not op
        for other in (*one.loads, *one.stores)
    )

File: src/optimize/loopmotion.rs
//! Sink loop stores nothing inside the loop observes to its single exit.
//!
//! Port of `qbopt/optimize/loopmotion.py`.

// Its callers live in transform.py, not yet ported.

use std::rc::Rc;
use std::collections::{BTreeMap, BTreeSet, HashMap};

use indexmap::IndexMap;

use crate::analysis::consts::{self, Cells, Known};
use crate::analysis::loops::{self, Loop};
use crate::analysis::ranges::{self, Interval};
use crate::analysis::regions::{self, RegionError, RegionLayout};
use crate::analysis::{effects, induction};
use crate::model::mir::{Arg, Const, Held, Kind, MemRef, MirBlock, MirBody, Op, Value};
use crate::objectfile::module::Space;
use crate::support::pyset::PySet;

/// Python's `bounds` dict, as the region layout that carries it.
type Bounds = IndexMap<(Space, i64), Vec<i64>>;

/// `id(op)`.
fn id(op: &Op) -> usize {
    std::ptr::from_ref(op) as usize
}

fn layout(bounds: Option<&Bounds>) -> Option<RegionLayout> {
    bounds.map(|bounds| RegionLayout {
        shared_segments: None,
        landmarks: bounds.iter().map(|(key, marks)| (*key, marks.clone())).collect(),
    })
}

/// `mir.overlapping(one, other, dgroup, bounds, known, other_known)`.
fn overlapping(
    one: &MemRef,
    other: &MemRef,
    bounds: Option<&Bounds>,
    known: Option<&BTreeMap<Value, Interval>>,
    other_known: Option<&BTreeMap<Value, Interval>>,
) -> Result<bool, String> {
    regions::overlapping(one, other, known, other_known, layout(bounds).as_ref())
        .map_err(|error: RegionError| format!("{error:?}"))
}

pub fn sunk_stores(
    body: &Rc<MirBody>,
    dgroup: &BTreeSet<i64>,
    bounds: Option<&Bounds>,
    handles_errors: bool,
) -> Result<Rc<MirBody>, String> {
    let mut body = body.clone();
    let predecessors = loops::predecessors(&body.blocks);
    for loop_ in loops::loops(&body.blocks, Some(body.entry)) {
        let scoped = ranges::bounded(&body)?;
        // With constants, as hoist asks: a store through the literal selector
        // 0A000h otherwise observes every frame and descriptor cell.
        let constant: BTreeMap<Value, Interval> = ranges::constants(&body, Some(dgroup), None).into_iter().collect();
        let mut intervals = HashMap::<usize, BTreeMap<Value, Interval>>::new();
        for block in &body.blocks {
            let mut here = constant.clone();
            if let Some(found) = scoped.get(&block.at) {
                here.extend(found.iter().map(|(value, interval)| (*value, interval.clone())));
            }
            for op in &block.ops {
                intervals.insert(id(op), here.clone());
            }
        }
        let blocks = body
            .blocks
            .iter()
            .map(|block| (block.at, block))
            .collect::<BTreeMap<i64, &MirBlock>>();
        // Python iterates `loop.body`, a frozenset, and the order reaches `moved`.
        let inside = loop_
            .body
            .iter()
            .copied()
            .collect::<PySet<i64>>()
            .iter()
            .map(|at| blocks[at])
            .collect::<Vec<_>>();
        let mut exits = Vec::new();
        for block in &inside {
            for to in &block.succ {
                if !loop_.body.contains(to) && !exits.contains(&(block.at, *to)) {
                    exits.push((block.at, *to));
                }
            }
        }
        if exits.len() != 1 {
            continue;
        }
        let (source, destination) = exits[0];
        if !blocks.contains_key(&destination) || predecessors[&destination] != BTreeSet::from([source]) {
            continue;
        }
        if inside.iter().any(|block| block.succ.is_empty()) {
            continue;
        }
        let operations = inside
            .iter()
            .flat_map(|block| block.ops.iter())
            .collect::<Vec<&Op>>();
        if operations.iter().any(|op| {
            effects::exposes_memory(op, handles_errors)
                || op.barrier()
                || matches!(
                    op.kind,
                    Kind::Call | Kind::Arg | Kind::Opaque | Kind::Escape | Kind::Return
                )
        }) {
            continue;
        }
        let exit_block = blocks[&destination];
        if exit_block.ops.is_empty() {
            continue;
        }
        let dominators = loops::dominators(&body.blocks, Some(body.entry));
        let empty = BTreeSet::new();
        let header_dominators = dominators.get(&loop_.header).unwrap_or(&empty);
        let address_values = body
            .blocks
            .iter()
            .filter(|block| !loop_.body.contains(&block.at) && header_dominators.contains(&block.at))
            .flat_map(|block| {
                block
                    .phis
                    .iter()
                    .map(|phi| phi.result)
                    .chain(block.ops.iter().flat_map(|op| op.defines.iter().copied()))
            })
            .collect::<BTreeSet<Value>>();
        let mut moved = Vec::<&Op>::new();
        for op in &blocks[&source].ops {
            if _unobserved(op, &operations, dgroup, bounds, Some(&intervals), &address_values)? {
                moved.push(op);
            }
        }
        let mut relocated = moved
            .iter()
            .map(|op| (id(op), (*op).clone()))
            .collect::<HashMap<usize, Op>>();
        if loop_.latches.len() == 1 && source == loop_.header {
            let latch = *loop_.latches.iter().next().expect("one latch");
            let outside = predecessors[&source]
                .difference(&loop_.body)
                .copied()
                .collect::<Vec<_>>();
            if outside.len() == 1 && latch != source && blocks[&latch].succ == [source] {
                let entry = outside[0];
                let nonempty = induction::nonempty(&body, &loop_);
                let invariant = if nonempty {
                    induction::invariant(&body, &loop_.body)
                } else {
                    BTreeSet::new()
                };
                for op in operations.iter().copied() {
                    if !relocated.contains_key(&id(op))
                        && _unobserved(op, &operations, dgroup, bounds, Some(&intervals), &address_values)?
                    {
                        let mut value = _exit_value(op, blocks[&source], blocks[&entry], latch, &body, dgroup, bounds, handles_errors)?
                            .map(Arg::Held);
                        if value.is_none() && blocks[&latch].ops.iter().any(|one| std::ptr::eq(op, one)) {
                            value = _invariant_value(op, &invariant, nonempty);
                            if value.is_none() {
                                value = _last_counter_value(op, &body, &loop_).map(Arg::Const);
                            }
                        }
                        if let Some(value) = value {
                            moved.push(op);
                            let mut changed = op.clone();
                            changed.args = op
                                .args
                                .iter()
                                .map(|arg| if matches!(arg, Arg::Held(_)) { value.clone() } else { arg.clone() })
                                .collect();
                            let mut uses = Vec::new();
                            if let Arg::Held(held) = &value {
                                uses.push(held.value);
                            }
                            for reference in &op.stores {
                                uses.extend([reference.base, reference.segment].into_iter().flatten());
                            }
                            let mut unique = Vec::new();
                            for one in uses {
                                if !unique.contains(&one) {
                                    unique.push(one);
                                }
                            }
                            changed.uses = unique;
                            relocated.insert(id(op), changed);
                        }
                    }
                }
            }
        }
        if moved.is_empty() {
            continue;
        }
        let identities = moved.iter().map(|op| id(op)).collect::<BTreeSet<usize>>();
        let mut updates = IndexMap::<i64, MirBlock>::new();
        for block in &inside {
            let mut changed = (*block).clone();
            changed.ops = block
                .ops
                .iter()
                .filter(|op| !identities.contains(&id(op)))
                .cloned()
                .collect();
            updates.insert(block.at, changed);
        }
        let mut changed = exit_block.clone();
        let anchor = exit_block.ops[0].at;
        changed.ops = moved
            .iter()
            .map(|op| Op {
                at: anchor,
                ..relocated[&id(op)].clone()
            })
            .chain(exit_block.ops.iter().cloned())
            .collect();
        updates.insert(destination, changed);
        let blocks = body
            .blocks
            .iter()
            .map(|block| updates.get(&block.at).cloned().unwrap_or_else(|| block.clone()))
            .collect();
        body = Rc::new(MirBody { blocks, ..MirBody::clone(&body) });
    }
    Ok(body)
}

fn _last_counter_value(op: &Op, body: &Rc<MirBody>, loop_: &Loop) -> Option<Const> {
    let [Arg::Held(stored)] = op.args.as_slice() else {
        return None;
    };
    let counter = induction::basics(body, loop_).get(&stored.value.id)?.clone();
    if counter.start.width() != stored.width || op.stores[0].width != stored.width {
        return None;
    }
    let last = induction::_last_counter(body, loop_, &counter, &consts::known(body, None, None, None, None), stored.width)?;
    Some(Const::new(consts::masked(&last, stored.width), stored.width))
}

fn _invariant_value(op: &Op, invariant: &BTreeSet<u32>, nonempty: bool) -> Option<Arg> {
    if !nonempty || op.args.len() != 1 {
        return None;
    }
    match &op.args[0] {
        Arg::Const(value) => Some(Arg::Const(value.clone())),
        Arg::Held(arg) if invariant.contains(&arg.value.id) => Some(Arg::Held(*arg)),
        _ => None,
    }
}

/// What `_exit_value` closes over.
struct _Exit<'a> {
    body: &'a MirBody,
    dgroup: &'a BTreeSet<i64>,
    bounds: Option<&'a Bounds>,
    reference: &'a MemRef,
    definitions: BTreeMap<Value, &'a Op>,
    blocks: BTreeMap<i64, &'a MirBlock>,
    predecessors: BTreeMap<i64, BTreeSet<i64>>,
    memory: Option<(IndexMap<(i64, usize), Cells>, IndexMap<i64, String>)>,
    handles_errors: bool,
}

impl _Exit<'_> {
    fn root(&self, mut arg: Arg) -> Arg {
        let mut seen = BTreeSet::new();
        while let Arg::Held(held) = &arg {
            if seen.contains(&held.value) {
                break;
            }
            seen.insert(held.value);
            let Some(defining) = self.definitions.get(&held.value) else {
                break;
            };
            if defining.kind != Kind::Copy || defining.args.len() != 1 {
                break;
            }
            if !defining
                .results
                .iter()
                .any(|result| matches!(result, Arg::Held(result) if result.value == held.value && result.width >= held.width))
            {
                break;
            }
            let width = held.width;
            arg = match &defining.args[0] {
                Arg::Held(source) if source.width >= width => Arg::Held(Held {
                    value: source.value,
                    width,
                }),
                Arg::Const(source) if source.width >= width => Arg::Const(Const::new(
                    &source.n & ((num_bigint::BigInt::from(1) << (width * 8)) - 1),
                    width,
                )),
                _ => break,
            };
        }
        arg
    }

    fn memory(&mut self) -> Result<&(IndexMap<(i64, usize), Cells>, IndexMap<i64, String>), String> {
        if self.memory.is_none() {
            let barriers = self
                .body
                .blocks
                .iter()
                .flat_map(|block| block.ops.iter())
                .filter(|one| one.barrier() || matches!(one.kind, Kind::Call | Kind::Escape | Kind::Opaque))
                .map(|one| (one.at, String::new()))
                .collect::<IndexMap<i64, String>>();
            let cells = consts::cells(self.body, self.dgroup, &barriers, None, None, None, None, None);
            self.memory = Some((cells, barriers));
        }
        Ok(self.memory.as_ref().expect("computed"))
    }

    fn stored_at(&mut self, at: i64, expected: Arg, active: &[(i64, Arg)]) -> Result<bool, String> {
        let expected = self.root(expected);
        let key = (at, expected.clone());
        if active.contains(&key) {
            return Ok(true); // inductive backedge; every entry path still needs a matching store
        }
        let block = self.blocks[&at];
        for index in (0..block.ops.len()).rev() {
            let previous = &block.ops[index];
            // A call is an operation like any other where its memory effects
            // are complete: its stores say what it can write.
            let opaque = previous.kind == Kind::Call && !previous.memory_complete;
            if previous.barrier() || opaque || matches!(previous.kind, Kind::Escape | Kind::Opaque) {
                return Ok(false);
            }
            if effects::exposes_memory(previous, self.handles_errors) {
                return Ok(false);
            }
            let mut overlaps = false;
            for written in &previous.stores {
                if overlapping(self.reference, written, self.bounds, None, None)? {
                    overlaps = true;
                    break;
                }
            }
            if overlaps {
                if let Arg::Const(expected) = &expected {
                    let mut fact = consts::initialized(previous, self.reference);
                    if fact.is_none() {
                        let (body, dgroup) = (self.body, self.dgroup);
                        let (facts, barriers) = self.memory()?;
                        let before = facts.get(&(at, index)).cloned().unwrap_or_default();
                        let nothing = IndexMap::new();
                        let mut asked = consts::memory_queries(body, &nothing, dgroup);
                        let after = consts::_kills(
                            &before,
                            previous,
                            &nothing,
                            dgroup,
                            barriers,
                            None,
                            None,
                            false,
                            Some(&mut asked),
                        );
                        fact = consts::_cell(&after, self.reference);
                    }
                    if fact == Some(Known::new(consts::masked(&expected.n, expected.width), expected.width)) {
                        return Ok(true);
                    }
                }
                let args = previous
                    .args
                    .iter()
                    .filter(|arg| matches!(arg, Arg::Const(_) | Arg::Held(_)))
                    .collect::<Vec<_>>();
                return Ok(previous.kind == Kind::Store
                    && previous.stores == [self.reference.clone()]
                    && args.len() == 1
                    && self.root(args[0].clone()) == expected);
            }
        }
        if at == self.body.entry || self.predecessors[&at].is_empty() {
            return Ok(false);
        }
        let phi = block
            .phis
            .iter()
            .find(|phi| matches!(&expected, Arg::Held(expected) if phi.result == expected.value));
        if let Some(phi) = phi {
            if phi.incoming.keys().copied().collect::<BTreeSet<_>>() != self.predecessors[&at] {
                return Ok(false);
            }
        }
        let mut active = active.to_vec();
        active.push(key);
        for parent in self.predecessors[&at].clone() {
            let next = match (phi, &expected) {
                (Some(phi), Arg::Held(expected)) => Arg::Held(Held {
                    value: *phi.incoming.get(&parent).expect("phi covers every predecessor"),
                    width: expected.width,
                }),
                _ => expected.clone(),
            };
            if !self.stored_at(parent, next, &active)? {
                return Ok(false);
            }
        }
        Ok(true)
    }
}

fn _exit_value(
    op: &Op,
    header: &MirBlock,
    entry: &MirBlock,
    latch: i64,
    body: &MirBody,
    dgroup: &BTreeSet<i64>,
    bounds: Option<&Bounds>,
    handles_errors: bool,
) -> Result<Option<Held>, String> {
    let values = op
        .args
        .iter()
        .filter_map(|arg| match arg {
            Arg::Held(held) => Some(*held),
            _ => None,
        })
        .collect::<Vec<_>>();
    let [stored] = values.as_slice() else {
        return Ok(None);
    };
    let definitions = body
        .blocks
        .iter()
        .flat_map(|block| block.ops.iter())
        .flat_map(|operation| operation.defines.iter().map(move |value| (*value, operation)))
        .collect::<BTreeMap<Value, &Op>>();
    let mut exit = _Exit {
        body,
        dgroup,
        bounds,
        reference: &op.stores[0],
        definitions,
        blocks: body.blocks.iter().map(|block| (block.at, block)).collect(),
        predecessors: loops::predecessors(&body.blocks),
        memory: None,
        handles_errors,
    };

    for phi in &header.phis {
        if phi.incoming.keys().copied().collect::<BTreeSet<_>>() != BTreeSet::from([entry.at, latch]) {
            continue;
        }
        let seed = exit.root(Arg::Held(Held {
            value: phi.incoming.get(&entry.at).copied().expect("incoming from entry"),
            width: stored.width,
        }));
        if exit.stored_at(entry.at, seed, &[])?
            && exit.stored_at(
                latch,
                Arg::Held(Held {
                    value: phi.incoming.get(&latch).copied().expect("incoming from latch"),
                    width: stored.width,
                }),
                &[],
            )?
        {
            return Ok(Some(Held {
                value: phi.result,
                width: stored.width,
            }));
        }
    }
    Ok(None)
}

fn _unobserved(
    op: &Op,
    operations: &[&Op],
    dgroup: &BTreeSet<i64>,
    bounds: Option<&Bounds>,
    intervals: Option<&HashMap<usize, BTreeMap<Value, Interval>>>,
    address_values: &BTreeSet<Value>,
) -> Result<bool, String> {
    let _ = dgroup;
    if op.kind != Kind::Store || !op.loads.is_empty() || !op.defines.is_empty() || op.stores.len() != 1 {
        return Ok(false);
    }
    let reference = &op.stores[0];
    if reference
        .addr
        .is_none_or(|addr| !matches!(addr.space, Space::Segment | Space::Frame))
        || reference.segment.is_some()
    {
        return Ok(false);
    }
    if reference
        .base
        .is_some_and(|base| !address_values.contains(&base) || reference.excludes.is_empty())
    {
        return Ok(false);
    }
    let known = intervals.and_then(|intervals| intervals.get(&id(op)));
    for one in operations {
        if std::ptr::eq(*one, op) {
            continue;
        }
        let other_known = intervals.and_then(|intervals| intervals.get(&id(one)));
        for other in one.loads.iter().chain(one.stores.iter()) {
            if overlapping(reference, other, bounds, known, other_known)? {
                return Ok(false);
            }
        }
    }
    Ok(true)
}

#[cfg(test)]
#[path = "loopmotion_tests.rs"]
mod tests;
